Fix for #13498

eaplatanios opened the pull request:
@skye I'm kind of indecisive it turns out. :P This is just your fix along with an equivalent fix for `RemoveEdge`. I don't know if you want to merge it given that you might make changes to your fix, but I'm putting here just in case. The `RemoveEdge` fix is pretty much equivalent and since the method is not used in the graph constructor, I didn't change its call sites to use `false` for the update node def argument. I'm not sure if that collides with anything else, but I guess the tests can tell us something (I didn't run them on my laptop because they take forever and I can't use it in the meantime).

eaplatanios committed:
Merge pull request #12 from tensorflow/master

Merge upstream changes.

eaplatanios committed:
Merge pull request #19 from tensorflow/master

Merge upstream changes.

eaplatanios committed:
Added Skye's fix plus a fix for "RemoveEdge".

tensorflow-jenkins commented:
Can one of the admins verify this patch?

skye commented:
Let's hold off on this until my patch is actually submitted :)

eaplatanios committed:
Fixed a problem with "RemoveEdge".

eaplatanios commented:
Sounds good! By the way, I just pushed a small update for `RemoveEdge`, because the previous version caused some problems.

caisq commented:
@tensorflow-jenkins test this please

eaplatanios committed:
Merge pull request #20 from tensorflow/master

Merge upstream changes.

eaplatanios committed:
Resolved merge conflicts.

eaplatanios committed:
Merge branch 'tensorflow-master' into node_def_fix

skye commented:
@eaplatanios you need to rebase. The AddControlEdge API changed during code review: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/graph/graph.h#L446

eaplatanios commented:
@skye I think I resolved this earlier today. Are there any changes I omitted?

On Oct 18, 2017, 2:29 PM -0400, Skye Wanderman-Milne <notifications@github.com>, wrote:
> @eaplatanios you need to rebase. The AddControlEdge API changed during code review: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/graph/graph.h#L446
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

eaplatanios committed:
Small fix.

skye commented:
Sorry I deleted that previous comment, I was somehow looking at the wrong thing.

skye commented on tensorflow/core/graph/graph.cc:
```diff
@@ -453,6 +453,20 @@ const Edge* Graph::AddControlEdge(Node* source, Node* dest,
   return AddEdge(source, kControlSlot, dest, kControlSlot);
 }
 
+void Graph::RemoveControlEdge(const Edge* e, bool update_node_def) {
```
Do you have need for update_node_def=false? If not, let's just remove it. The contract with AddControLEdge() is that it will make sure the NodeDef is in sync with the added control edge (unless allow_duplicates=true, which is only in one weird specific case that we would ideally change). So we can assume that we should continue to keep the NodeDef in sync here.

eaplatanios replied:
Yeah I only added that option in order to be consistent with your previous change because I was unsure of where else it might be necessary to not keep the NodeDef in sync. I'll remove that.

skye commented on tensorflow/core/graph/graph_constructor.cc:
```diff
@@ -840,7 +840,7 @@ Status GraphConstructor::Convert() {
             EdgeInfo(inputs[i].name, inputs[i].index, node, i));
       } else if (inputs[i].index == Graph::kControlSlot) {
-        g_->AddControlEdge(inputs[i].node, node);
+        g_->AddControlEdge(inputs[i].node, node, /*update_node_def=*/false);
```
Do you need to update this file? Also unit tests for your new method?

eaplatanios replied:
These were just changes you had in your early draft implementation and I copied them over. I'll remove them and check that everything works fine. Also, I'll add tests for the new method once I get the chance this week. :)

eaplatanios commented:
No worries. It was actually super well timed. I like pushed the changes a few minutes before you sent me that. :)

On Oct 18, 2017, 2:44 PM -0400, Skye Wanderman-Milne <notifications@github.com>, wrote:
> Sorry I deleted that previous comment, I was somehow looking at the wrong thing.
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

eaplatanios committed:
Requested edits.

eaplatanios committed:
Minor edit in the documentation.

eaplatanios committed:
Added a unit test for the "RemoveControlEdge" method.

eaplatanios commented:
@skye I added a simple test for the new method. I'm not sure what your convention is for unit tests. Please let me know if you want me to test for some specific condition other than the simple case I added. :)

skye commented on tensorflow/core/graph/graph_test.cc:
```diff
@@ -477,6 +491,44 @@ TEST_F(GraphTest, AddControlEdge) {
+  ASSERT_TRUE(edge_1 != nullptr);
+  ASSERT_TRUE(edge_2 != nullptr);
+
+  const Edge* edge_1_obtained = FindInputControlEdge(a, edge_1);
```
You can remove L509-521, as this is testing AddControlEdge, which is already tested above. I would then change FindInputControlEdge to "bool ControlEdgeExists()".

skye commented on tensorflow/core/graph/graph_test.cc:
```diff
@@ -477,6 +491,44 @@ TEST_F(GraphTest, AddControlEdge) {
+  EXPECT_EQ(edge_2_obtained->dst_input(), Graph::kControlSlot);
+
+  graph_.RemoveControlEdge(edge_1);
+  ASSERT_TRUE(FindInputControlEdge(a, edge_1) == nullptr);
```
Here or possibly in the helper function you should check that the NodeDef doens't contain the control input either.

skye commented on tensorflow/core/graph/graph_test.cc:
```diff
@@ -477,6 +491,44 @@ TEST_F(GraphTest, AddControlEdge) {
+
+  graph_.RemoveControlEdge(edge_2);
+  ASSERT_TRUE(FindInputControlEdge(a, edge_1) == nullptr);
+  ASSERT_TRUE(FindInputControlEdge(b, edge_2) == nullptr);
```
Test removing a duplicate control edge

eaplatanios committed:
Made the requested changes.

eaplatanios commented:
@skye I made all of the suggested changes. :)

skye commented on tensorflow/core/graph/graph.cc:
```diff
@@ -453,6 +453,20 @@ const Edge* Graph::AddControlEdge(Node* source, Node* dest,
+    e->dst_->MaybeCopyOnWrite();
+    auto* inputs = e->dst_->props_->node_def.mutable_input();
+    for (auto it = inputs->begin(); it != inputs->end(); ++it) {
+      if (*it == strings::StrCat("^", e->src_->name())) {
```
nit: pull the strings::StrCat out of the for loop so you're not computing it every time (you could even not construct a string at all, but it's fine)

eaplatanios replied:
Good catch. :)

skye commented on tensorflow/core/graph/graph.h:
```diff
@@ -451,6 +451,10 @@ class Graph {
   // REQUIRES: The edge must exist.
   void RemoveEdge(const Edge* edge);
 
+  // Removes control edge `edge` from the graph.
```
Add that this updates the node def

skye commented on tensorflow/core/graph/graph_test.cc:
```diff
@@ -118,6 +118,24 @@ class GraphTest : public ::testing::Test {
     LOG(FATAL) << name;
   }
 
+  bool ControlEdgeExistsInGraphOrNodeDef(const Node* src,
```
This logic isn't quite right... you want to test that both the Edge* and the NodeDef input were removed, not either. Maybe divide this into two functions, or change it to "void ExceptControlEdgeRemoved" and it does both checks, or something.

eaplatanios replied:
This checks if it exists in either the graph or the NodeDef, meaning that it returns `false` only if it's been removed from both. Isn't that what we want?

skye replied:
Ah right, my bad.

skye commented on tensorflow/core/graph/graph_test.cc:
```diff
@@ -477,6 +495,42 @@ TEST_F(GraphTest, AddControlEdge) {
+
+  // Test removing a duplicate control edge.
+  const Edge* edge_3 = graph_.AddControlEdge(c, a);
+  const Edge* edge_4 = graph_.AddControlEdge(c, a);
```
Sorry, I was unclear about this... this won't actually add a duplicate edge unless you pass "allow_duplicates" here. This will be kind of a weird test case (I don't actually expect anyone to use allow_duplicates=true along with RemoveControlEdge), but I think it will be good to at least document the behavior in this test.

eaplatanios replied:
Good point. I'll add a comment for this.

eaplatanios committed:
Made the requested edits.

eaplatanios commented:
@skye I made the changes you suggested except the one about the boolean function which I think is correct, but could you please check and let me know if I'm confused about something? :)

eaplatanios committed:
Made a small requested change.

skye commented on tensorflow/core/graph/graph_test.cc:
```diff
@@ -477,6 +496,47 @@ TEST_F(GraphTest, AddControlEdge) {
+  // will not be added. That's why we expect edge_4 to be a null
+  // pointer. We are not testing with allow_duplicates set to true,
+  // as that is a highly unlikely use case that does not make much
+  // sense.
```
I actually meant to test with allow_duplicates = true. But this is fine.

eaplatanios replied:
I can add that too if you want. :)

skye replied:
I need to run soon, and will be gone until next week, so this is fine, or do it in a separate PR and I'll review when I get back.

eaplatanios commented:
Sounds good! Have fun wherever you’re going! :)

On Oct 18, 2017, 8:57 PM -0400, Skye Wanderman-Milne <notifications@github.com>, wrote:
> @skye commented on this pull request.
> In tensorflow/core/graph/graph_test.cc:
> > +  ASSERT_TRUE(ControlEdgeExistsInGraphOrNodeDef(a, b));
> +
> +  graph_.RemoveControlEdge(edge_1);
> +  ASSERT_TRUE(!ControlEdgeExistsInGraphOrNodeDef(c, a));
> +  ASSERT_TRUE(ControlEdgeExistsInGraphOrNodeDef(a, b));
> +
> +  graph_.RemoveControlEdge(edge_2);
> +  ASSERT_TRUE(!ControlEdgeExistsInGraphOrNodeDef(c, a));
> +  ASSERT_TRUE(!ControlEdgeExistsInGraphOrNodeDef(a, b));
> +
> +  // Test removing a duplicate control edge.
> +  // Note that unless allow_duplicates is true, the duplicate edge
> +  // will not be added. That's why we expect edge_4 to be a null
> +  // pointer. We are not testing with allow_duplicates set to true,
> +  // as that is a highly unlikely use case that does not make much
> +  // sense.
> I need to run soon, and will be gone until next week, so this is fine, or do it in a separate PR and I'll review when I get back.
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

eaplatanios committed:
Merge pull request #23 from tensorflow/master

Merge upstream changes.

gunan commented:
Jenkins, test this please.

gunan commented:
Jenkins, test this please.
